### PR TITLE
Return ConnectionFailureException which contains the connection state info in onConnectionFailure (Part II)

### PR DESCRIPTION
**Pull request by yufengwangca**
This PR is to enhance the onConnectionFailure method of the GetConnectedDeviceCallback interface to provide more context about the failure, specifically by introducing a new ConnectionFailure exception subclass that includes a ConnectionState enum. This approach can indeed provide valuable debugging and metrics data, as well as improve the user experience by providing more detailed information about the connection failure.

Testing with java-matter-controller with modified all-clusters-app.

### CODE REVIEW

File: src/app/CASESessionManager.cpp [hunk 1]
```diff
@@ -30,7 +30,8 @@ CHIP_ERROR CASESessionManager::Init(chip::System::Layer * systemLayer, const CAS
 }
 
 void CASESessionManager::FindOrEstablishSession(const ScopedNodeId & peerId, Callback::Callback<OnDeviceConnected> * onConnection,
-                                                Callback::Callback<OnDeviceConnectionFailure> * onFailure
+                                                Callback::Callback<OnDeviceConnectionFailure> * onFailure,
+                                                Callback::Callback<OnExtendedDeviceConnectionFailure> * onExtendedConnectionFailure
```
Comment (bzbarsky-apple) [inline src/app/CASESessionManager.cpp:34]: I believe @tcarmelveilleux was suggesting we have a new FindOrEstablishSession with a struct argument instead of continuing to add arguments, and make the old API call that.

In any case, this is breaking the existing API, which we should not be doing.  We should be adding another FindOrEstablishSession overload instead.

File: src/app/CASESessionManager.cpp [hunk 2]
```diff
@@ -54,6 +55,14 @@ void CASESessionManager::FindOrEstablishSession(const ScopedNodeId & peerId, Cal
             {
                 onFailure->mCall(onFailure->mContext, peerId, CHIP_ERROR_NO_MEMORY);
             }
+
+            if (onExtendedConnectionFailure != nullptr)
```
Comment (bzbarsky-apple) [inline src/app/CASESessionManager.cpp:59]: I don't think we should have a way to provide both onFailure and onExtendedConnectionFailure.  One or the other, but not both.

File: src/app/CASESessionManager.cpp [hunk 3]
```diff
@@ -54,6 +55,14 @@ void CASESessionManager::FindOrEstablishSession(const ScopedNodeId & peerId, Cal
...
+            {
+                // Initialize the ExtendedConnectionFailureInfo object
+                ExtendedConnectionFailureInfo failureInfo(peerId, CHIP_ERROR_NO_MEMORY, CaseConnectionState::Uninitialized);
```
Comment (bzbarsky-apple) [inline src/app/CASESessionManager.cpp:62]: Just call this struct ConnnectionFailureInfo.

File: src/app/OperationalSessionSetup.cpp [hunk 4]
```diff
@@ -46,24 +46,24 @@ using chip::AddressResolve::ResolveResult;
 
 namespace chip {
 
-void OperationalSessionSetup::MoveToState(State aTargetState)
+void OperationalSessionSetup::MoveToState(CaseConnectionState aTargetState)
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.cpp:49]: Please do a separate PR under this one to rename the state.  That will make reviews much simpler.

File: src/app/CASESessionManager.cpp [hunk 5]
```diff
@@ -54,6 +55,14 @@ void CASESessionManager::FindOrEstablishSession(const ScopedNodeId & peerId, Cal
             {
```
Comment (gmarcosb) [inline src/app/CASESessionManager.cpp:55]: I think it could be worthwhile to do a conditional here rather than calling both, i.e. if (onExtendedCOnnectionFailure != null) {} else { existing code }

I think also treating this as a oneof, i.e. not letting callers pass both, is worthwhile

File: src/app/OperationalSessionSetup.cpp [hunk 6]
```diff
@@ -339,6 +348,23 @@ void OperationalSessionSetup::NotifyConnectionCallbacks(Cancelable & failureRead
...
+        VerifyOrDie(!performingAddressUpdate);
+        Callback::Callback<OnExtendedDeviceConnectionFailure> * cb =
+            Callback::Callback<OnExtendedDeviceConnectionFailure>::FromCancelable(extendedFailureReady.mNext);
+
+        cb->Cancel();
+
+        if (error != CHIP_NO_ERROR)
+        {
+            // Initialize the ExtendedConnectionFailureInfo object
+            ExtendedConnectionFailureInfo failureInfo(peerId, error, state);
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.cpp:363]: OK, so let's pick a concrete failure case.  Say we got an address, then we could not send Sigma1 to it.  We will have state == ResolvingAddress here.  Is that what API consumers expect?

Similarly, if we resolve a thing and then send Sigma1 and then the CASE handshake fails, we will have state == NeedsAddress at this point.  I am 100% sure that's not what consumers expect in terms of getting actionable data.

Basically, our current internal state after failure does not track "what failed" very well.

File: src/controller/java/src/chip/devicecontroller/ConnectionFailureException.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    super(errorCode, message);
+    this.connectionState = mapIntToConnectionState(connectionState);
+  }
+
+  public ConnectionState getConnectionState() {
+    return connectionState;
+  }
+
+  // Helper method to map an int to ConnectionState enum
+  private ConnectionState mapIntToConnectionState(int value) {
```
Comment (gmarcosb) [inline src/controller/java/src/chip/devicecontroller/ConnectionFailureException.java:48]: I think this would be easier to maintain if it's kept in the enum, see e.g. https://github.com/project-chip/connectedhomeip/blob/15f54da67d08f40b18d366433c492b7c1c8dd89a/src/controller/java/src/chip/onboardingpayload/CommissioningFlow.kt#L24

File: src/app/OperationalSessionSetup.h [hunk 8]
```diff
@@ -181,16 +209,17 @@ class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate,
...
      * If the session already exists, `onConnection` will be called immediately,
      * before the Connect call returns.
      *
-     * `onFailure` may be called before the Connect call returns, for error
+     * `onFailure` and `onExtendedConnectionFailure` may be called before the Connect call returns, for error
      * cases that are detected synchronously (e.g. inability to start an address
      * lookup).
      */
-    void Connect(Callback::Callback<OnDeviceConnected> * onConnection, Callback::Callback<OnDeviceConnectionFailure> * onFailure);
+    void Connect(Callback::Callback<OnDeviceConnected> * onConnection, Callback::Callback<OnDeviceConnectionFailure> * onFailure,
```
Comment (bzbarsky-apple) [inline src/app/OperationalSessionSetup.h:221]: Again, this is changing a public API.

File: src/app/ReadClient.cpp [hunk 9]
```diff
@@ -1229,7 +1229,7 @@ CHIP_ERROR ReadClient::EstablishSessionToPeer()
     ChipLogProgress(DataManagement, "Trying to establish a CASE session for subscription");
     auto * caseSessionManager = InteractionModelEngine::GetInstance()->GetCASESessionManager();
     VerifyOrReturnError(caseSessionManager != nullptr, CHIP_ERROR_INCORRECT_STATE);
-    caseSessionManager->FindOrEstablishSession(mPeer, &mOnConnectedCallback, &mOnConnectionFailureCallback);
+    caseSessionManager->FindOrEstablishSession(mPeer, &mOnConnectedCallback, &mOnConnectionFailureCallback, nullptr);
```
Comment (bzbarsky-apple) [inline src/app/ReadClient.cpp:1232]: See, none of these changes outside CASESessionManager and OperatonalSessionSetup should be needed if we were not breaking the API.

File: src/app/OperationalSessionSetup.h [hunk 10]
```diff
@@ -109,6 +109,29 @@ class OperationalDeviceProxy : public DeviceProxy
...
+    SecureConnected,  // CASE session established.
+    WaitingForRetry,  // No address known, but a retry is pending.  Added at
+                      // end to make logs easier to understand.
+};
+
+struct ExtendedConnectionFailureInfo
+{
+    const ScopedNodeId & peerId;
+    CHIP_ERROR error;
+    CaseConnectionState connectionState;
```
Comment (gmarcosb) [inline src/app/OperationalSessionSetup.h:128]: Can we also get [CaseSession::state](https://github.com/project-chip/connectedhomeip/blob/15f54da67d08f40b18d366433c492b7c1c8dd89a/src/protocols/secure_channel/CASESession.h#L194) & an equivalent for PaseSession (seems like [mNextExpectedMessage](http://google3/third_party/connectedhomeip/next/src/protocols/secure_channel/PASESession.h), but maybe in a more readable format) in here?

**Review by gmarcosb — COMMENTED** [review]
Thank you for doing this Yufeng!

File: src/app/OperationalSessionSetup.h [hunk 11]
```diff
@@ -109,6 +109,29 @@ class OperationalDeviceProxy : public DeviceProxy
     ScopedNodeId mPeerScopedNodeId;
 };
 
+enum class CaseConnectionState : uint8_t
```
Comment (pidarped) [inline src/app/OperationalSessionSetup.h:112]: It seems we are using the term 'Connection' interchangeably with 'Session Establishment'. The API is FindOrEstablishSession() but herein we are calling it a connection. This would, especially, muddy the waters when we have TCP as the underlying transport. Would have preferred if it were named as Session. Changing it everywhere would have a lot of touch points though.

File: src/app/OperationalSessionSetup.h [hunk 12]
```diff
@@ -109,6 +109,29 @@ class OperationalDeviceProxy : public DeviceProxy
...
+{
+    Uninitialized,    // Error state: OperationalSessionSetup is useless
+    NeedsAddress,     // No address known, lookup not started yet.
+    ResolvingAddress, // Address lookup in progress.
+    HasAddress,       // Have an address, CASE handshake not started yet.
+    Connecting,       // CASE handshake in progress.
```
Comment (pidarped) [inline src/app/OperationalSessionSetup.h:118]:
```suggestion
    SessionEstablishing,       // CASE handshake in progress.
```

File: src/app/OperationalSessionSetup.h [hunk 13]
```diff
@@ -109,6 +109,29 @@ class OperationalDeviceProxy : public DeviceProxy
...
+    SecureConnected,  // CASE session established.
```
Comment (pidarped) [inline src/app/OperationalSessionSetup.h:119]:
```suggestion
    SessionEstablished,  // CASE session established.
```

File: src/app/OperationalSessionSetup.h (hunk 11 above)
Comment (pidarped) [inline src/app/OperationalSessionSetup.h:112]:
```suggestion
enum class CaseSessionState : uint8_t
```

File: src/app/CASESessionManager.cpp (hunk 5 above)
Comment (yufengwangca) [reply src/app/CASESessionManager.cpp:55]: Acked

File: src/controller/java/src/chip/devicecontroller/ConnectionFailureException.java (hunk 7 above)
Comment (yufengwangca) [reply src/controller/java/src/chip/devicecontroller/ConnectionFailureException.java:48]: Yes, that would be an ideal solution. However, we do face limitations because this class is generated from the native side, which only supports primary types as input parameters.

File: src/app/OperationalSessionSetup.h (hunk 10 above)
Comment (yufengwangca) [reply src/app/OperationalSessionSetup.h:128]: In the context of this PR, it all about establishing a CASE (Conditional Action for System Exceptions) for handling the onConnectionFailure method.  In order to manage the state for PASE (Preserve and Safeguard Exception Handling), we must approach it through a different API and within a separate PR.

File: src/app/OperationalSessionSetup.h (hunk 11 above)
Comment (yufengwangca) [reply src/app/OperationalSessionSetup.h:112]: I agree with this renaming, but it changed the public API and we need to update all references. Track this change in https://github.com/project-chip/connectedhomeip/issues/29303

File: src/app/OperationalSessionSetup.h (hunk 12 above)
Comment (yufengwangca) [reply src/app/OperationalSessionSetup.h:118]: https://github.com/project-chip/connectedhomeip/issues/29303

File: src/app/OperationalSessionSetup.h (hunk 13 above)
Comment (yufengwangca) [reply src/app/OperationalSessionSetup.h:119]: https://github.com/project-chip/connectedhomeip/issues/29303

File: src/app/OperationalSessionSetup.cpp (hunk 4 above)
Comment (yufengwangca) [reply src/app/OperationalSessionSetup.cpp]: https://github.com/project-chip/connectedhomeip/pull/29305

File: src/app/OperationalSessionSetup.cpp (hunk 6 above)
Comment (yufengwangca) [reply src/app/OperationalSessionSetup.cpp]: The state provides valuable context about the error, pinpointing the precise conditions under which the failure occurs.

This information is crucial for our statistical analysis, as it allows us to track the occurrence of errors within our test matrix. Therefore, it's essential that we log the state simultaneously when recording the error.

I will transfer all the comprehensive modifications for OperationalSessionSetup to https://github.com/project-chip/connectedhomeip/pull/29305, which encompasses the fundamental alterations for this Feature Request (FR). Once we reach a consensus on OperationalSessionSetup, this pull request should undergo a swift review.

Comment (bzbarsky-apple) [reply src/app/OperationalSessionSetup.cpp]: > pinpointing the precise conditions under which the failure occurs.

My point, which I keep reapeating, is that it ***does nothing of the sort***.  At least not given how https://github.com/project-chip/connectedhomeip/pull/29305 is using the state.

File: src/controller/java/src/chip/devicecontroller/ConnectionFailureException.java (hunk 7 above)
Comment (gmarcosb) [reply src/controller/java/src/chip/devicecontroller/ConnectionFailureException.java:48]: Apologies I think you may have missed what I was saying - yes, from JNI, you'll need mapIntToConneciton state, but you'll want instead something like this:

```
 public enum ConnectionState {
    private final int number;
    public int getNumber() { return number }

    private ConnectionState(int number) { this.number = number }
    UNINITIALIZED(0 // Error state: OperationalSessionSetup is useless
    NEEDS_ADDRESS(1), // No address known, lookup not started yet.
...
  }
```

Then, here, you'd simply do:

```
for (state in ConnectionState.values()) {
   if (state.getNumber == value) {
      return state;
  }
}
```

(You can optimize with a map if necessary, but as this is a small list with a constrained size, for loop should do the trick)

File: src/app/OperationalSessionSetup.h [hunk 14]
```diff
@@ -155,6 +155,31 @@ typedef void (*OnDeviceConnectionRetry)(void * context, const ScopedNodeId & pee
 class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate, public AddressResolve::NodeListener
```
Comment (gmarcosb) [inline src/app/OperationalSessionSetup.h:155]: So as this extends `SessionEstablishmentDelegate`, can we not extend that delegate to have a method (with a default-implemented {}) which is something like `connectionStateUpdated`, which passes the connection state, called on every state change?

**Comment by yufengwangca** [comment]
Will post a new one based on https://github.com/project-chip/connectedhomeip/pull/29305
